source/differentiation/ad/ad_drivers.cc

```diff
@@ -0,0 +1,2260 @@
+      // because a non-existant tape_index has been used as an argument, or
+      // because the tape exists but has not been used (in a way that
+      // initiated a status update). For example, on can create a tape in one
+      // section of code and then query the status of all existing tapes in a
+      // completely different section of code that knows nothing about the
+      // first tape. There is no prerequisite that the first tape is ever
+      // used, and it can therefore not have a status. So, in this case
+      // there's not much we can do other than to report that the tape does
+      // not require retaping.
+      if (it_status_tape == status.end())
+        return false;
+
+      const auto status_tape = it_status_tape->second;
+
+      // See ADOL-C manual section 1.7 and comments in last paragraph of
+      // section 3.1
+      Assert(status_tape < 4 && status_tape >= -2,
+             ExcIndexRange(status_tape, -2, 4));
+      return (status_tape < 0);
+    }
+
+
+    template <typename ADNumberType>
+    bool
+    TapedDrivers<
+      ADNumberType,
+      double,
+      typename std::enable_if<ADNumberTraits<ADNumberType>::type_code ==
+                              NumberTypes::adolc_taped>::type>::
+      last_action_requires_retaping() const
+    {
+      return requires_retaping(active_tape);
+    }
+
+
+    template <typename ADNumberType>
+    void
+    TapedDrivers<
+      ADNumberType,
+      double,
+      typename std::enable_if<ADNumberTraits<ADNumberType>::type_code ==
+                              NumberTypes::adolc_taped>::type>::
+      remove_tape(const typename Types<ADNumberType>::tape_index tape_index)
+    {
+      Assert(is_registered_tape(tape_index),
+             ExcMessage(
+               "This tape does not exist, and therefore cannot be cleared."));
+      removeTape(tape_index, TapeRemovalType::ADOLC_REMOVE_COMPLETELY);
+    }
+
+
+    template <typename ADNumberType>
+    void
+    TapedDrivers<
+      ADNumberType,
+      double,
+      typename std::enable_if<ADNumberTraits<ADNumberType>::type_code ==
+                              NumberTypes::adolc_taped>::type>::
+      reset(const bool clear_registered_tapes)
+    {
+      active_tape       = Numbers<ADNumberType>::invalid_tape_index;
+      is_recording_flag = false;
+      status.clear();
+      if (clear_registered_tapes)
+        {
+          const std::vector<typename Types<ADNumberType>::tape_index>
+            registered_tape_indices = get_registered_tape_indices();
+          for (const auto &tape_index : registered_tape_indices)
+            remove_tape(tape_index);
+        }
+    }
+
+
+    template <typename ADNumberType>
+    void
+    TapedDrivers<ADNumberType,
+                 double,
+                 typename std::enable_if<
+                   ADNumberTraits<ADNumberType>::type_code ==
+                   NumberTypes::adolc_taped>::type>::print(std::ostream &stream)
```

Comment:
sigh :-( This indentation is entirely unreadable...

Reply:
Yes, this is slightly unfortunate :-(